src/main/kotlin/minesweeper/domain/cell/Cell.kt

```diff
@@ -1,38 +1,93 @@
-import minesweeper.domain.common.rem
+import minesweeper.domain.board.BoardStatus
+import minesweeper.domain.board.MineMaker
 
-sealed class Cell(val position: Position)
+sealed class Cell(val position: Position) {
+    private var state: CellStatus = CellStatus.CLOSE
 
-class Cells private constructor(
-    private val cells: List<Cell>,
-    val mineIndices: NumberSet
-) : List<Cell> by cells {
+    fun open(): Cell {
+        check(state != CellStatus.OPEN) { "cell (${position.x}, ${position.y}) was already opened." }
+        state = CellStatus.OPEN
+        return this
+    }
+
+    fun isClosed() = this.state == CellStatus.CLOSE
+
+    fun isOpen() = this.state == CellStatus.OPEN
+}
+
+class Cells(private val cells: List<Cell>) : List<Cell> by cells {
+
+    fun open(position: Position): BoardStatus {
+        val cell = findCell(position).open()
+        if (cell !is Empty) {
+            return BoardStatus.BOOM
+        }
+        cell.openNearbyCells()
+        return BoardStatus.SAFE
+    }
+
+    fun hasClosedEmptyCell() = cells.any { it is Empty && it.isClosed() }
+
+    private fun Empty.openNearbyCells() {
+        if (this.numberOfNearbyMines != 0) {
+            return
+        }
+        getNearbyCells(this).forEach { nearbyCell ->
+            nearbyCell
+                .takeIf { nearbyCell is Empty && nearbyCell.isClosed() }
+                ?.let { open(nearbyCell.position) }
+        }
+    }
+
+    fun getNearbyCells(cell: Cell): Cells {
+        val width = last().position.x + 1
+        val height = last().position.y + 1
+        val nearbyPositions = getNearbyPositions(cell, width, height)
+
+        return Cells(nearbyPositions.map { findCell(it) })
+    }
+
+    private fun getNearbyPositions(
+        cell: Cell,
+        width: Int,
+        height: Int,
+    ) = NearbyDirection.values().mapNotNull {
+        val nearbyX = cell.position.x + it.x
+        val nearbyY = cell.position.y + it.y
+        val nearbyIndex = nearbyY * width + nearbyX
+
+        if (nearbyX.isBetweenRange(width) && nearbyY.isBetweenRange(height) && nearbyIndex < size) {
+            Position(nearbyX, nearbyY)
+        } else null
+    }
+
+    private fun findCell(position: Position) =
+        cells.find { it.position == position } ?: throw IllegalArgumentException("no cells found in that position.")
+
+    private fun Int.isBetweenRange(limit: Int) = this in 0 until limit
 
     companion object {
-        fun of(width: PositiveInt, height: PositiveInt, numberOfMines: PositiveInt, mineStrategy: MineStrategy): Cells {
-            val numberOfCells = width * height
-            val mineIndices = mineStrategy.getMineIndices(numberOfCells, numberOfMines)
-            val size = (width * height).value
+        fun of(width: Int, height: Int, numberOfMines: Int, mineMaker: MineMaker): Cells {
+            val cells = initCells(width, height)
+            val mineCells = mineMaker.createMines(width, height, numberOfMines)
```

Comment:
```suggestion
            val mineCells = mineMaker.createMines(cells, numberOfMines)
```

만들어진 cells와 만들어져야하는 mine 개수를 넣어서 지뢰들을 가지고 오는게 좋지 않을까요?

Reply:
이건 아래 코멘트와도 연관된 질문이 있는데요!

`mineCells + mineCell의 위치를 제외한 Cells의 List를 만드는게` 와 같이 코드를 개선해서 현재 로직은 mine 을 먼저 생성한 후, 마인이 없는 위치에만 empty 를 생성하도록 변경했는데요~

mineMaker.createMines(cells, numberOfMines) 와 같이 변경하게 되면 현재는 width, height 값을 어디에서도 관리하지 않기 때문에 MineMaker 에서도 cells 만으로는 보드의 width와 height을 알 수 없습니다. 따라서, 보드의 크기 내에서 랜덤 포지션을 만들 수가 없다고 생각되는데요. 
랜덤 포지션을 생성하는 로직에는 반드시 width, height 값이 필요하다고 생각됩니다!

Reply:
그렇군요! 제가 해당 부분을 제시한 이유는 이미 만들어진 cells에 대해서 suffle 하여 앞에서부터 마인의 숫자로 자르면 더 괜찮지 않을까 하여, 추천드렸습니다!! 이게 아니면 미리 만들어 놓은 position을 가지고도 할 수 있어 보여요 😊